### ISSUE TITLE
memory issue, prepare_model_for_kbit_training

### ISSUE THREAD

**Issue by wenzhifang** [bitsandbytes-foundation/bitsandbytes#1569]
I am using quantized base model (NF4) and do LoRA fine-tuning.
When i call prepare_model_for_kbit_training to wrap the model, the memory consumption is even higher than the bf16 counterpart.

**Comment by alceballosa** [comment]
Having the the same issue when using a non-quantized base model and trying to finetune with QLORA int4, getting the following mem comsumption (Ministral 8B, seq length 4096, bsz 1), using the following bnb config:

`        bnb_config = BitsAndBytesConfig(
             load_in_4bit=True,
             bnb_4bit_use_double_quant=True,
             bnb_4bit_quant_type="nf4",
             bnb_4bit_compute_dtype=torch_dtype,
         )
`

bf16 with plain LORA: 72840 MB 
quantized to int4 for QLORA but not using prepare_model_for_kbit_training: 63082 MB
quantized to int4 for QLORA and using prepare_model_for_kbit_training:  81076 MB

I'm assuming that omitting the prepare_model_for_kbit_training call will cause silent errors during training, but I'm a bit confused about these results. What's the recommended thing to do in this scenario?

Thanks in advance for your time!


Versions:

peft -> 0.15.2
torch ->  2.6.0+cu126
transformers -> 4.51.3
bitsandbytes -> 0.45.5

**Comment by matthewdouglas** [comment]
Hi folks,
`prepare_model_for_kbit_training` does cast some parameters to `float32`. I'm afraid I'm not 100% positive on the reasoning and the impact to training results. @BenjaminBossan is this something you could explain?

**Comment by alceballosa** [comment]
Thanks for the prompt reply @matthewdouglas . If that's the case, I suppose it might be a matter of "savings at scale" since I'm using a single GPU and bsz=1. I might try to run the same pipeline with a smaller model and a larger batch size. Would be great to get some confirmation from Benjamin on whether the use of the preparation function is mandatory whenever we're using 4 or 8 bit quantization.

**Comment by BenjaminBossan** [comment]
Sorry for the late reply, I somehow missed this thread. AFAIK, the reason why `prepare_model_for_kbit_training` casts adapter weights to float32 is that it was found to improve fine-tuning performance while only slightly increasing memory requirements. I doubt that there is a significant improvement in each and every use case, so definitely try how it works without. `prepare_model_for_kbit_training` also does some work around gradient checkpointing. If you need that, you'd have to make those calls manually.

**Comment by TimDettmers** [comment]
Closing this issue. The PEFT maintainer explained in the comments that `prepare_model_for_kbit_training` intentionally casts adapter weights to float32, which increases memory usage. This is by design to improve fine-tuning performance. For lower memory usage, you can skip `prepare_model_for_kbit_training` or manually set `model.gradient_checkpointing_enable()` and cast specific layers to the precision you need.
